### ISSUE TITLE
print_uninitialized model segfaults at runtime with stanc3

### ISSUE THREAD

**Issue by rok-cesnovar** [stan-dev/stanc3#309]
The following model when compiled with stanc3 fails at runtime:

```
transformed data {
  real udummy;
  { 
    real v;
    real vv[10];
    vector[12] vvv;
    matrix[10,10] vvvv;
    matrix[3,4] vvvvv[5];
    print("transformed data: ",v,"  ",vv[1],"  ",vvv[2],"  ",vvvv[3,4],"  ",vvvvv[1,2,3]);
  }
}
parameters {
  real y;
}
transformed parameters {
  real wdummy;
  { 
    real w;
    real ww[10];
    vector[12] www;
    matrix[10,10] wwww;
    matrix[3,4] wwwww[5];
    print("transformed parameters: ",w," ",ww[1]," ",www[2]," ",wwww[3,4]," ",wwwww[1,2,3]);
  }
}
model {
  real z;
  real zz[10];
  vector[12] zzz;
  matrix[10,10] zzzz;
  matrix[10,10] zzzzz[10];
  print("model: ",z," ",zz[1]," ",zzz[2]," ",zzzz[3,4]," ",zzzzz[3,4,5]); 
  y ~ normal(0,1);
}
```
It runs OK with the old compiler. This test is part of the test/interface/ test in cmdstan.

Attaching code from both the old compiler and the new compiler:

[print_uninitialized_stanc2.hpp.txt](https://github.com/stan-dev/stanc3/files/3665882/print_uninitialized_stanc2.hpp.txt)
[print_uninitialized_stanc3.hpp.txt](https://github.com/stan-dev/stanc3/files/3665883/print_uninitialized_stanc3.hpp.txt)

**Comment by rok-cesnovar** [comment]
The minimal example this can be reproduced is: 

```
parameters {
  real y;
}
model {
  real z;
  print("model: ",z); 
  y ~ normal(0,1);
}
```
This model however works fine: 

```
parameters {  
  real z;
}
model {
  print("model: ",z);
}
```

**Comment by seantalts** [comment]
Thanks for the report and the minimal example! I'll look into it hopefully today or tomorrow :)

**Comment by seantalts** [comment]
Hmm, I'm thinking we should actually just print an error if this occurs. @bob-carpenter or @VMatthijs what do you think about that idea?

For reference, I think Stan 2 just initializes it to a quiet NaN, which means it can get pretty far before throwing an exception.

**Comment by seantalts** [comment]
(And @rybern recently just added a check for uninitialized variables, but it's not turned on by default).

**Comment by seantalts** [comment]
Fixed by #326
